{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "239d2480",
   "metadata": {},
   "source": [
    "# Quality Check: Assemblies\n",
    "\n",
    "\n",
    "Step 0: Checking to make sure you have your assemblies\n",
    "\n",
    "Step 1: Running Quast on on the assembly\n",
    "\n",
    "Step 2: Running CheckM on the assembly"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b3fa9c1",
   "metadata": {},
   "source": [
    "## Getting Started\n",
    "\n",
    "You will need to rerun this section each time you come back to this notebook to reset all directories and variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e9373313",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set the variables for your netid and xfile\n",
    "id = \"YOUR_ID\"\n",
    "accessions = \"MY_ACCESSIONS\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3d3c9105",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set directories\n",
    "xfile_dir = \"/my_dir_path/\" + id + \"/00_getting_data\"\n",
    "work_dir = \"/my_dir_path/\" + id + \"/04_assembly_qc\"\n",
    "data_dir = \"/my_dir_path/\" + id + \"/00_getting_data\"\n",
    "unicycler_dir = \"/my_dir_path/\" + id + \"/03_assembly\" + \"/out_unicycler\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f65dacd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Go into the working directory\n",
    "%cd $work_dir"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68869704",
   "metadata": {},
   "source": [
    "## Creating a config file\n",
    "The scripts below executes code that requires certain variables to be set. So we don't need to edit the code in the script, we are going to use a config file that defines all of these variables for us. Then when we want to use these variables in the script, we will \"source\" the config file to set the variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "958665ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a config file with all of the variables you need\n",
    "# notice that we are using the reads post-trimming, and post-human removal\n",
    "!echo \"export ID=$id\" > config.sh\n",
    "!echo \"export ACESSIONS=$accessions\" >> config.sh\n",
    "!echo \"export WORK_DIR=$work_dir\" >> config.sh\n",
    "!echo \"export DATA_DIR=$data_dir\" >> config.sh\n",
    "!echo \"export UNICYCLER_DIR=$unicycler_dir\" >> config.sh\n",
    "!echo \"export QUAST=/path_to_containers/quast:5.2.0--py39pl5321h4e691d4_3.sif\" >> config.sh\n",
    "!echo \"export CHECKM=/path_to_containers/checkm2\\:1.0.1--pyh7cba7a3_0.sif\" >> config.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6633e4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check the config file to be sure it is correct\n",
    "!cat config.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f659efe",
   "metadata": {},
   "source": [
    "## Step 0:  Checking the contig count\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9560d189",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check that we have contigs after the assembly\n",
    "import os\n",
    "list = data_dir + '/' + accessions\n",
    "print(list)\n",
    "lines = open(list).read().splitlines()\n",
    "for file in lines:\n",
    "    command = 'egrep \">\" ' + unicycler_dir + '/' + file + '/assembly.fasta | wc -l'\n",
    "    os.system(command)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5afd05f0",
   "metadata": {},
   "source": [
    "Great! Looks like we have all of our contig files from unicycler (or if not, then check the slurm logs to fix). "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4903cdc1",
   "metadata": {},
   "source": [
    "## Step 1: Quast\n",
    "\n",
    "How good are our assemblies? We can check the quality by running tools that look at the contigs produced by our assembly algorithms. \n",
    "\n",
    "Let's see what the quality of our assemblies from unicycler, using a bioinformatics tool called quast. We can run this tool on multiple assemblies at once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e1df048d",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_code = '''#!/bin/bash\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --nodes=1             \n",
    "#SBATCH --time=12:00:00   \n",
    "#SBATCH --partition=standard\n",
    "#SBATCH --account=your_account\n",
    "#SBATCH --array=0-15                         \n",
    "#SBATCH --output=04A_quast-%a.out\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH --mem-per-cpu=5G                                    \n",
    "\n",
    "pwd; hostname; date\n",
    "\n",
    "source ./config.sh\n",
    "names=($(cat $DATA_DIR/$ACCESSIONS))\n",
    "\n",
    "SAMPLE_ID=${names[${SLURM_ARRAY_TASK_ID}]}\n",
    "\n",
    "### create output directories for the reports\n",
    "### note that we are going to compare both assemblies at once\n",
    "OUTDIR=${WORK_DIR}/out_quast\n",
    "\n",
    "### create the outdir if it does not exist\n",
    "if [[ ! -d \"$OUTDIR\" ]]; then\n",
    "  echo \"$OUTDIR does not exist. Directory created\"\n",
    "  mkdir $OUTDIR\n",
    "fi\n",
    "\n",
    "### Contigs to use\n",
    "CONTIGS=$UNICYCLER_DIR/${SAMPLE_ID}/assembly.fasta\n",
    "\n",
    "### Run Quast\n",
    "apptainer run ${QUAST} quast -t 24 \\\n",
    "        -o $OUTDIR/${SAMPLE_ID} \\\n",
    "        -m 500 \\\n",
    "        $CONTIGS\n",
    "'''\n",
    "\n",
    "with open('04A_quast.sh', mode='w') as file:\n",
    "    file.write(my_code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3814e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's run the sbatch script, this should take 10-15 minutes to run once the job starts\n",
    "# You can go on to Step 2 in the meantime.\n",
    "# The quality reports for quast and checkm can run at the same time\n",
    "!sbatch ./04A_quast.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2474106e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if quast is running\n",
    "!squeue --user=$id"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa743720",
   "metadata": {},
   "source": [
    "## Step 2: Checkm2\n",
    "\n",
    "Checkm2 is another tool that allows you to produce a quality report on the assembled contigs.\n",
    "\n",
    "The documentation can be found [here](https://github.com/chklovski/CheckM2)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76539c71",
   "metadata": {},
   "source": [
    "### Checkm2 database file\n",
    "\n",
    "This tool requires a database file to run. More information on downloading the database can be found in the documentation. The current database has been downloaded and saved in the following location:\n",
    "\n",
    "/path-to_databases/checkm2_database/uniref100.KO.1.dmnd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2532f289",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a script to run on the contigs\n",
    "# A few important points:\n",
    "my_code = '''#!/bin/bash\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --nodes=1             \n",
    "#SBATCH --time=24:00:00   \n",
    "#SBATCH --partition=standard\n",
    "#SBATCH --account=your_account\n",
    "#SBATCH --array=0-15                      \n",
    "#SBATCH --output=04B_checkm-%a.out\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH --mem-per-cpu=5G                                    \n",
    "\n",
    "pwd; hostname; date\n",
    "\n",
    "source ./config.sh\n",
    "names=($(cat $DATA_DIR/$ACCESSIONS))\n",
    "\n",
    "SAMPLE_ID=${names[${SLURM_ARRAY_TASK_ID}]}\n",
    "\n",
    "### create output directory for the report\n",
    "OUTDIR=${WORK_DIR}/out_checkm\n",
    "\n",
    "### create the outdirs if they do not exist\n",
    "if [[ ! -d \"$CHECKM_OUTDIR\" ]]; then\n",
    "  echo \"$CHECKM_OUTDIR does not exist. Directory created\"\n",
    "  mkdir -p $CHECKM_OUTDIR\n",
    "fi\n",
    "\n",
    "### Run checkm\n",
    "apptainer run ${CHECKM} checkm2 \\\n",
    "        predict --threads 24 \\\n",
    "        --input $UNICYCLER_DIR \\\n",
    "        -x fasta \\\n",
    "        --output-directory $OUTDIR/${SAMPLE_ID} \\\n",
    "        --database_path /groups/bhurwitz/databases/checkm2_database/uniref100.KO.1.dmnd  \n",
    "'''\n",
    "\n",
    "with open('04B_checkm.sh', mode='w') as file:\n",
    "    file.write(my_code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b24186e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's run the sbatch script, this should take ~1 hour to run\n",
    "!sbatch ./04B_checkm.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f70dfab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Welcome back, let's see if the job is still running\n",
    "!squeue --user=$id"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69a24a6d",
   "metadata": {},
   "source": [
    "#### Let's check out the assembly stats from Checkm2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef0a005f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd $work_dir/out_checkm\n",
    "!cat */quality_report.tsv #cats all reports"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
